expression/builtin.go

```diff
@@ -117,16 +118,67 @@ func CheckIllegalMixCollation(funcName string, args []Expression, evalType types
 	if len(args) < 2 {
 		return nil
 	}
-	_, _, coercibility, legal := inferCollation(args...)
+	_, dstCharset, coercibility, legal := inferCollation(args...)
 	if !legal {
 		return illegalMixCollationErr(funcName, args)
 	}
 	if coercibility == CoercibilityNone && evalType != types.ETString {
 		return illegalMixCollationErr(funcName, args)
 	}
+
+	// For every constant and string type arguments, MySQL will try to convert its charset to the inferred charset,
+	// return illegalMixCollationErr if convert failed.
+	// e.g.
+	// 		select _utf8 'a' collate utf8_general_ci = '😁'; # error
+	// 		select _utf8 'a' collate utf8_general_ci = 'ㅂ'; # fine
+	// it is because '😁' cannot convert to utf8mb3, it is both fine in tidb for now, and I would not change the
+	// behavior because we tread utf8mb3 same as utf8mb4 except insert, just keep things simple.
+	// Since we only have utf8mb4 and its subset charset for now, so that just check the data's validity is ok, no
+	// need to convert it.
+	// In the future, we may add more charsets support, but it actually still utf8mb4 encoding at runtime, so we
+	// also no need to do the convert.
+	for _, arg := range args {
+		_, ok := arg.(*Constant)
+		if ok && arg.GetType().EvalType() == types.ETString {
+			val, isNull, err := arg.EvalString(nil, chunk.Row{})
+			if err != nil || isNull {
+				return err
+			}
+
+			if !isValidString(val, arg.GetType().Charset, dstCharset) {
+				return illegalMixCollationErr(funcName, args)
+			}
+		}
+	}
+
 	return nil
 }
 
+// isValidString test if the val is valid for charset dstChs.
+// TODO: Remove this function when the full charset is supported, we will have a better way to do this job.
+func isValidString(val string, srcChs, dstChs string) bool {
+	if srcChs == dstChs {
+		return true
+	}
+	switch dstChs {
+	case charset.CharsetASCII:
+		for _, c := range val {
+			if c >= 0x80 {
```

Comment:
Is this logic covered by tests? Could we add one?

Reply:
the test is added, PTAL